src/Compilers/CSharp/Portable/CodeGen/CodeGenerator_RefSafety.cs

```diff
@@ -0,0 +1,131 @@
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeGen;
+
+internal partial class CodeGenerator
+{
+    private static bool MightEscapeTemporaryRefs(BoundCall node, bool used, AddressKind? receiverAddressKind)
+    {
+        return MightEscapeTemporaryRefs(
+            used: used,
+            returnType: node.Type,
+            returnRefKind: node.Method.RefKind,
+            receiverType: !node.Method.RequiresInstanceReceiver ? null : node.ReceiverOpt?.Type,
+            receiverScope: node.Method.TryGetThisParameter(out var thisParameter) ? thisParameter?.EffectiveScope : null,
+            receiverAddressKind: receiverAddressKind,
+            isReceiverReadOnly: node.Method.IsEffectivelyReadOnly,
+            parameters: node.Method.Parameters);
+    }
+
+    private static bool MightEscapeTemporaryRefs(BoundObjectCreationExpression node, bool used)
+    {
+        return MightEscapeTemporaryRefs(
+            used: used,
+            returnType: node.Type,
+            returnRefKind: RefKind.None,
+            receiverType: null,
+            receiverScope: null,
+            receiverAddressKind: null,
+            isReceiverReadOnly: false,
+            parameters: node.Constructor.Parameters);
+    }
+
+    private static bool MightEscapeTemporaryRefs(BoundFunctionPointerInvocation node, bool used)
+    {
+        FunctionPointerMethodSymbol method = node.FunctionPointer.Signature;
+        return MightEscapeTemporaryRefs(
+            used: used,
+            returnType: node.Type,
+            returnRefKind: method.RefKind,
+            receiverType: null,
+            receiverScope: null,
+            receiverAddressKind: null,
+            isReceiverReadOnly: false,
+            parameters: method.Parameters);
+    }
+
+    private static bool MightEscapeTemporaryRefs(
+        bool used,
+        TypeSymbol returnType,
+        RefKind returnRefKind,
+        TypeSymbol? receiverType,
+        ScopedKind? receiverScope,
+        AddressKind? receiverAddressKind,
+        bool isReceiverReadOnly,
+        ImmutableArray<ParameterSymbol> parameters)
+    {
+        Debug.Assert(receiverAddressKind is null || receiverType is not null);
+
+        // We check the signature of the method, counting potential `ref` sources and destinations
+        // to determine whether a `ref` can be captured by the method.
+        // The emit layer then uses this information to avoid reusing temporaries that are passed by ref to such methods.
+
+        // number of outputs that can capture `ref`s
+        int refTargets = 0;
+        // number of inputs that can contain `ref`s
+        int refSources = 0;
+
+        if (used && (returnRefKind != RefKind.None || returnType.IsRefLikeOrAllowsRefLikeType()))
+        {
+            // If returning by ref or returning a ref struct, the result might capture `ref`s.
+            refTargets++;
+        }
+
+        if (receiverType is not null)
+        {
+            receiverScope ??= ScopedKind.None;
```

Comment:
>??= [](http://example.com/codeflow?start=26&length=3)

Do we ever have a receiver and `receiverScope == null`?

Reply:
Failing CI shows there are actually cases like that:

```
Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics.NativeIntegerTests.BinaryOperators
Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics.NativeIntegerTests.ReadOnlyField_VirtualMethods
Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics.NativeIntegerTests.Retargeting_06
Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics.NativeIntegerTests.SwitchStatement_02
Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics.NativeIntegerTests.SwitchStatement_01
Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics.NativeIntegerTests.MultipleTypeRefs_01
```

It fails on `nint.ToString` which is a `NativeIntegerMethodSymbol` that does not implement TryGetThisParameter. It seems that the correct fix is overriding that, so I will do that.